impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnsh/builtin/BuiltinShellCommandRegistryFactory.java

```diff
@@ -75,15 +75,14 @@ private BuiltinShellCommandRegistry(LookupContext shellContext) {
             this.mavenParser = new MavenParser();
             this.shellEncryptInvoker = new EncryptInvoker(shellContext.invokerRequest.lookup(), contextCopier());
             this.encryptParser = new EncryptParser();
-            Set<Command> commands = new HashSet<>(EnumSet.allOf(Command.class));
-            Map<Command, String> commandName = new HashMap<>();
-            Map<Command, CommandMethods> commandExecute = new HashMap<>();
-            for (Command c : commands) {
-                commandName.put(c, c.name().toLowerCase());
-            }
-            commandExecute.put(Command.MVN, new CommandMethods(this::mvn, this::mvnCompleter));
-            commandExecute.put(Command.MVNENC, new CommandMethods(this::mvnenc, this::mvnencCompleter));
-            registerCommands(commandName, commandExecute);
+            Map<String, CommandMethods> commandExecute = new HashMap<>();
+            commandExecute.put("!", new CommandMethods(this::shell, this::defaultCompleter));
+            commandExecute.put("cd", new CommandMethods(this::cd, this::cdCompleter));
+            commandExecute.put("ls", new CommandMethods(this::ls, this::defaultCompleter));
```

Comment:
JLine [defines a bunch of commands](https://github.com/jline/jline3/blob/master/demo/src/main/java/org/apache/felix/gogo/jline/Posix.java) already.  They are currently leveraging Felix Gogo, but it should be fairly easy to abstract from it and make them reusable.  
Would that make sense to reuse them here ?

Reply:
I was intentionally limiting "command set" to "navigation" only. Keep core simple: can run mvn/mvnenc (like in real shell, provided by distro) and you can navigate your workdir. All the rest is not in maven core anyway, and IMO should be provided like extensions IMO. I think we don't want a super-uber generic script/shell here, just leave the door open for whoever wants it, be able to do it.

Reply:
Then, I'd remove `ls` and see if we can provide a posix set of commands in a separate extension.
I'd really like a property in `~/.m2/maven.properties` that could be parsed to discover extensions: this could include a list of GAvoids which would contain project-less mojos that could be run as commands, such as toolbox, etc...

Reply:
Removed `ls`, and yes, we may want to provide some standard extensions. And yes, that would be useful feat.